### PR TITLE
Merge miracl evaluator

### PR DESCRIPTION
**Pull request by jordiclive**
addressing #829

### CODE REVIEW

**Comment by KennethEnevoldsen** [comment]
Looks great so far, it is still a draft so I will wait to review it. Feel free to ping me when it is ready.

**Comment by jordiclive** [comment]
@KennethEnevoldsen Thanks, ready. I did some quick tests with a dummy dataset to check the same results as before.

File: mteb/evaluation/evaluators/RerankingEvaluator.py [hunk 1]
```diff
@@ -102,21 +104,106 @@ def compute_metrics_batched(self, model):
             all_query_flattened = [
                 q for sample in self.samples for q in sample["query"]
             ]
-            all_query_embs = self._encode_unique_texts(
-                all_query_flattened, encode_corpus_func
-            )
+            if self.evaluator_type == "standard":
+                all_query_embs = self._encode_unique_texts(
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/RerankingEvaluator.py:108]: It seems like we want to encode unique texts regardless of whether we use miracl or not

File: mteb/evaluation/evaluators/RerankingEvaluator.py [hunk 2]
```diff
@@ -102,21 +104,106 @@ def compute_metrics_batched(self, model):
...
+                all_conf_scores=all_conf_scores,
+            )
+        mean_ap = np.mean(all_ap_scores)
+        mean_mrr = np.mean(all_mrr_scores)
+
+        # Compute nAUCs
+        naucs_map = self.nAUC_scores(all_conf_scores, all_ap_scores, "map")
+        naucs_mrr = self.nAUC_scores(all_conf_scores, all_mrr_scores, "mrr")
+
+        return {**{"map": mean_ap, "mrr": mean_mrr}, **naucs_map, **naucs_mrr}
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/RerankingEvaluator.py:188]: I would probably move all the metric calculations out of the "_encode_candidates"

File: mteb/evaluation/evaluators/RerankingEvaluator.py [hunk 3]
```diff
@@ -192,23 +258,131 @@ def compute_metrics_individual(self, model):
...
+            docs_emb = all_docs_embs[docs_idx : docs_idx + num_doc]
+            docs_idx += num_doc
+
+            fake_qid = str(query_idx)
+            results[fake_qid] = self.rerank(query_emb, docs_emb)
+            qrels[fake_qid] = {
+                str(i): 1 if doc in positive else 0 for i, doc in enumerate(docs)
+            }
+
+        scores_miracl = self._collect_miracl_results(results, qrels)
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/RerankingEvaluator.py:319]: Is there a difference in the encode besides this step?

File: mteb/evaluation/evaluators/RerankingEvaluator.py [hunk 4]
```diff
@@ -102,21 +104,106 @@ def compute_metrics_batched(self, model):
...
+            results = self._encode_candidates(
+                encode_corpus_func=encode_corpus_func,
+                encode_queries_func=encode_queries_func,
+                batched=False,
+            )
+        elif self.evaluator_type == "miracl":
+            results = self._encode_candidates_miracl_individual(
+                encode_queries_func=encode_queries_func,
+                encode_corpus_func=encode_corpus_func,
+            )
```
Comment (KennethEnevoldsen) [inline mteb/evaluation/evaluators/RerankingEvaluator.py:155]: I believe we can unify the encode part here?

File: mteb/evaluation/evaluators/RerankingEvaluator.py (hunk 3 above)
Comment (jordiclive) [reply mteb/evaluation/evaluators/RerankingEvaluator.py:336]: a difference with the unbatched?

File: mteb/evaluation/evaluators/RerankingEvaluator.py (hunk 4 above)
Comment (jordiclive) [reply mteb/evaluation/evaluators/RerankingEvaluator.py]: @KennethEnevoldsen I'm not sure where without trying to dismantle the MIRACL code.

I confirmed that the results are the same as before this MR. I improved readability, by using the same code structure for both standard and MIRACL modes. I addressed two of your suggestions—moved metric calculations to a separate method and verified encode_unique_texts is indeed ok for MIRACL—and merged the upstream changes with the task name.

File: mteb/evaluation/evaluators/RerankingEvaluator.py (hunk 3 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/RerankingEvaluator.py:336]: Batched and unmatched suppose

File: mteb/evaluation/evaluators/RerankingEvaluator.py (hunk 4 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/RerankingEvaluator.py]: > @KennethEnevoldsen I'm not sure where without trying to dismantle the MIRACL code.

Right. Then let us keep it as is for now. It is already a great deal better than before.

File: mteb/evaluation/evaluators/RerankingEvaluator.py (hunk 3 above)
Comment (KennethEnevoldsen) [reply mteb/evaluation/evaluators/RerankingEvaluator.py:336]: Check, and it is not. I think we could unify these more, but will merge this PR in for now as it is already a good improvement

**Pull request merged**